File: models/prompt.py
"""Prompt 五要素数据模型 — 对齐 Seedance 2.0 导演法结构"""

import json
import time
import uuid


class PromptElement:
    """单个提示词要素"""

    def __init__(self, label: str, key: str, placeholder: str, value: str = ''):
        self.label = label
        self.key = key
        self.placeholder = placeholder
        self.value = value

    def to_dict(self) -> dict:
        return {
            'label': self.label,
            'key': self.key,
            'placeholder': self.placeholder,
            'value': self.value,
        }


class SeedancePrompt:
    """Seedance Prompt 五要素模型

    五要素导演法:
    1. 角色指定 (subject)  — 主体描述：外貌、服装、表情
    2. 场景设定 (scene)    — 环境、时代、光照
    3. 动作/剧情 (action)  — 运动序列，支持多步描述
    4. 镜头语言 (camera)   — 推/拉/摇/移/环绕/跟随
    5. 氛围/声音 (atmosphere) — 风格、色调、音效、BGM
    """

    ELEMENTS = [
        PromptElement(
            '角色指定', 'subject',
            '描述主体：外貌特征、服装、表情、姿态。\n例：一位穿着红色汉服的年轻女子，长发飘逸，眼神温柔'
        ),
        PromptElement(
            '场景设定', 'scene',
            '描述环境：时代背景、地点、光照条件。\n例：19世纪的伦敦大街，黄昏暖光，街道两旁有维多利亚式建筑'
        ),
        PromptElement(
            '动作/剧情', 'action',
            '描述运动序列：主体动作、互动、时间顺序。\n例：她缓缓转身，微笑着向镜头走来，裙摆随风飘动'
        ),
        PromptElement(
            '镜头语言', 'camera',
            '描述运镜方式：推/拉/摇/移/环绕/跟随/升降。\n例：慢推镜头，从全景逐渐推进到面部特写'
        ),
        PromptElement(
            '氛围/声音', 'atmosphere',
            '描述风格、色调、音效。\n例：电影感，暖色调，背景有轻微虚化，柔和的钢琴配乐'
        ),
    ]

    def __init__(self):
        self.id = str(uuid.uuid4())[:8]
        self.name = ''
        self.created_at = time.time()
        self.updated_at = time.time()

        # 五要素
        self.subject = ''
        self.scene = ''
        self.action = ''
        self.camera = ''
        self.atmosphere = ''

        # 视频参数
        self.task_type = 'text2video'
        self.model = 'doubao-seedance-2-0-260128'
        self.resolution = '720p'
        self.duration = 5
        self.ratio = '16:9'

        # 素材引用
        self.ref_assets = []  # [{'id': ..., 'role': 'reference_image'}, ...]

    def build_prompt_text(self) -> str:
        """将五要素组合为最终 Prompt 文本"""
        parts = []
        if self.subject:
            parts.append(self.subject)
        if self.scene:
            parts.append(self.scene)
        if self.action:
            parts.append(self.action)
        if self.camera:
            parts.append(self.camera)
        if self.atmosphere:
            parts.append(self.atmosphere)
        return '，'.join(parts) if parts else ''

    def to_dict(self) -> dict:
        """序列化为字典"""
        return {
            'id': self.id,
            'name': self.name,
            'created_at': self.created_at,
            'updated_at': self.updated_at,
            'subject': self.subject,
            'scene': self.scene,
            'action': self.action,
            'camera': self.camera,
            'atmosphere': self.atmosphere,
            'task_type': self.task_type,
            'model': self.model,
            'resolution': self.resolution,
            'duration': self.duration,
            'ratio': self.ratio,
            'ref_assets': self.ref_assets,
            'prompt_text': self.build_prompt_text(),
        }

    def to_api_payload(self) -> dict:
        """导出为 Seedance API 兼容的 JSON 格式"""
        payload = {
            'model': self.model,
            'content': [
                {'type': 'text', 'text': self.build_prompt_text()}
            ],
            'video_config': {
                'resolution': self.resolution,
                'duration': self.duration,
                'ratio': self.ratio,
                'fps': 24,
            }
        }
        # 添加素材引用
        for ref in self.ref_assets:
            payload['content'].append({
                'type': 'image_url' if ref.get('type') == 'image' else ref.get('type', 'image_url'),
                'image_url': {'url': ref.get('path', '')},
                'image_role': ref.get('role', 'reference_image'),
            })
        return payload

    @classmethod
    def from_dict(cls, data: dict) -> 'SeedancePrompt':
        """从字典反序列化"""
        prompt = cls()
        prompt.id = data.get('id', prompt.id)
        prompt.name = data.get('name', '')
        prompt.created_at = data.get('created_at', prompt.created_at)
        prompt.updated_at = data.get('updated_at', prompt.updated_at)
        prompt.subject = data.get('subject', '')
        prompt.scene = data.get('scene', '')
        prompt.action = data.get('action', '')
        prompt.camera = data.get('camera', '')
        prompt.atmosphere = data.get('atmosphere', '')
        prompt.task_type = data.get('task_type', 'text2video')
        prompt.model = data.get('model', 'doubao-seedance-2-0-260128')
        prompt.resolution = data.get('resolution', '720p')
        prompt.duration = data.get('duration', 5)
        prompt.ratio = data.get('ratio', '16:9')
        prompt.ref_assets = data.get('ref_assets', [])
        return prompt

    def save(self, filepath: str):
        """保存到 JSON 文件"""
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(self.to_dict(), f, ensure_ascii=False, indent=2)

    @classmethod
    def load(cls, filepath: str) -> 'SeedancePrompt':
        """从 JSON 文件加载"""
        with open(filepath, 'r', encoding='utf-8') as f:
            return cls.from_dict(json.load(f))

File: static/js/prompt.js
/* ═══════════════════════════════════════════════════════
   prompt.js — Prompt 编辑器交互逻辑
   ═══════════════════════════════════════════════════════ */

document.addEventListener('DOMContentLoaded', () => {
    const state = {
        currentProjectId: null,
        subject: '',
        scene: '',
        action: '',
        camera: '',
        atmosphere: '',
        task_type: 'text2video',
        model: 'doubao-seedance-2-0-260128',
        resolution: '720p',
        ratio: '16:9',
        duration: 5,
        ref_assets: [],
    };

    // ── DOM 元素 ─────────────────────────────────────
    const inputs = {
        subject: document.getElementById('input-subject'),
        scene: document.getElementById('input-scene'),
        action: document.getElementById('input-action'),
        camera: document.getElementById('input-camera'),
        atmosphere: document.getElementById('input-atmosphere'),
    };

    const params = {
        taskType: document.getElementById('param-task-type'),
        model: document.getElementById('param-model'),
        duration: document.getElementById('param-duration'),
        durationDisplay: document.getElementById('duration-display'),
    };

    const preview = {
        text: document.getElementById('preview-text'),
        json: document.getElementById('preview-json'),
        metaModel: document.getElementById('meta-model'),
        metaTask: document.getElementById('meta-task'),
        metaResolution: document.getElementById('meta-resolution'),
        metaRatio: document.getElementById('meta-ratio'),
        metaDuration: document.getElementById('meta-duration'),
    };

    // ── 五要素输入监听 ───────────────────────────────
    Object.keys(inputs).forEach(key => {
        inputs[key].addEventListener('input', () => {
            state[key] = inputs[key].value;
            updatePreview();
        });
    });

    // ── 参数监听 ─────────────────────────────────────
    params.taskType.addEventListener('change', () => {
        state.task_type = params.taskType.value;
        updatePreview();
    });

    params.model.addEventListener('change', () => {
        state.model = params.model.value;
        updatePreview();
    });

    params.duration.addEventListener('input', () => {
        state.duration = parseInt(params.duration.value);
        params.durationDisplay.textContent = state.duration;
        updatePreview();
    });

    // 分辨率 chips
    document.getElementById('param-resolution').addEventListener('click', (e) => {
        const chip = e.target.closest('.chip');
        if (!chip) return;
        document.querySelectorAll('#param-resolution .chip').forEach(c => c.classList.remove('active'));
        chip.classList.add('active');
        state.resolution = chip.dataset.value;
        updatePreview();
    });

    // 宽高比 chips
    document.getElementById('param-ratio').addEventListener('click', (e) => {
        const chip = e.target.closest('.chip');
        if (!chip) return;
        document.querySelectorAll('#param-ratio .chip').forEach(c => c.classList.remove('active'));
        chip.classList.add('active');
        state.ratio = chip.dataset.value;
        updatePreview();
    });

    // ── 模板加载 ─────────────────────────────────────
    document.getElementById('template-chips').addEventListener('click', async (e) => {
        const chip = e.target.closest('.chip');
        if (!chip) return;
        const templateId = chip.dataset.templateId;

        try {
            const template = await api.get(`/api/templates/${templateId}`);
            // 填充五要素
            ['subject', 'scene', 'action', 'camera', 'atmosphere'].forEach(key => {
                if (template[key]) {
                    state[key] = template[key];
                    inputs[key].value = template[key];
                }
            });
            // 填充参数
            if (template.task_type) {
                state.task_type = template.task_type;
                params.taskType.value = template.task_type;
            }
            if (template.resolution) {
                state.resolution = template.resolution;
                document.querySelectorAll('#param-resolution .chip').forEach(c => {
                    c.classList.toggle('active', c.dataset.value === template.resolution);
                });
            }
            if (template.ratio) {
                state.ratio = template.ratio;
                document.querySelectorAll('#param-ratio .chip').forEach(c => {
                    c.classList.toggle('active', c.dataset.value === template.ratio);
                });
            }
            if (template.duration) {
                state.duration = template.duration;
                params.duration.value = template.duration;
                params.durationDisplay.textContent = template.duration;
            }

            updatePreview();
            Toast.success(`已加载模板: ${template.name}`);
        } catch (err) {
            Toast.error(`加载模板失败: ${err.message}`);
        }
    });

    // ── 预览更新 ─────────────────────────────────────
    function updatePreview() {
        // 合成 Prompt 文本
        const parts = [];
        if (state.subject) parts.push(state.subject);
        if (state.scene) parts.push(state.scene);
        if (state.action) parts.push(state.action);
        if (state.camera) parts.push(state.camera);
        if (state.atmosphere) parts.push(state.atmosphere);

        const promptText = parts.join('，');

        if (promptText) {
            preview.text.textContent = promptText;
            preview.text.classList.remove('preview-placeholder');
        } else {
            preview.text.innerHTML = '<span class="preview-placeholder">在左侧填写五要素，此处将实时显示合成的 Prompt 文本...</span>';
        }

        // 参数汇总
        const modelNames = {
            'doubao-seedance-2-0-260128': 'Seedance 2.0',
            'doubao-seedance-1-5-pro': 'Seedance 1.5 Pro',
            'doubao-seedance-1-0-pro': 'Seedance 1.0 Pro',
            'doubao-seedance-1-0-pro-fast': 'Seedance 1.0 Pro Fast',
        };
        const taskNames = {
            'text2video': '文生视频',
            'image2video': '图生视频',
            'video_edit': '视频编辑',
            'video_extend': '视频延长',
        };

        preview.metaModel.textContent = modelNames[state.model] || state.model;
        preview.metaTask.textContent = taskNames[state.task_type] || state.task_type;
        preview.metaResolution.textContent = state.resolution;
        preview.metaRatio.textContent = state.ratio;
        preview.metaDuration.textContent = `${state.duration}秒`;

        // API JSON 预览
        const apiPayload = {
            model: state.model,
            content: [
                { type: 'text', text: promptText }
            ],
            video_config: {
                resolution: state.resolution,
                duration: state.duration,
                ratio: state.ratio,
                fps: 24,
            }
        };

        // 添加素材引用
        state.ref_assets.forEach(ref => {
            apiPayload.content.push({
                type: 'image_url',
                image_url: { url: ref.path || '' },
                image_role: ref.role || 'reference_image',
            });
        });

        preview.json.textContent = JSON.stringify(apiPayload, null, 2);
    }

    // ── 保存项目 ─────────────────────────────────────
    const saveModal = document.getElementById('save-modal');
    const saveNameInput = document.getElementById('save-project-name');

    document.getElementById('btn-save-project').addEventListener('click', () => {
        saveModal.style.display = 'flex';
        saveNameInput.value = state.name || '';
        saveNameInput.focus();
    });

    document.getElementById('btn-close-save').addEventListener('click', () => {
        saveModal.style.display = 'none';
    });

    document.getElementById('btn-cancel-save').addEventListener('click', () => {
        saveModal.style.display = 'none';
    });

    document.getElementById('btn-confirm-save').addEventListener('click', async () => {
        const name = saveNameInput.value.trim();
        if (!name) {
            Toast.error('请输入项目名称');
            return;
        }

        try {
            const data = { ...state, name };
            if (state.currentProjectId) data.id = state.currentProjectId;
            const result = await api.post('/api/prompts', data);
            state.currentProjectId = result.id;
            state.name = name;
            saveModal.style.display = 'none';
            Toast.success('项目保存成功');
            loadProjectsList();
        } catch (err) {
            Toast.error(`保存失败: ${err.message}`);
        }
    });

    // 回车保存
    saveNameInput.addEventListener('keydown', (e) => {
        if (e.key === 'Enter') document.getElementById('btn-confirm-save').click();
    });

    // ── 加载项目 ─────────────────────────────────────
    const projectsSection = document.getElementById('projects-section');
    const projectsList = document.getElementById('projects-list');

    document.getElementById('btn-load-project').addEventListener('click', () => {
        if (projectsSection.style.display === 'none') {
            projectsSection.style.display = 'block';
            loadProjectsList();
        } else {
            projectsSection.style.display = 'none';
        }
    });

    async function loadProjectsList() {
        try {
            const data = await api.get('/api/prompts');
            if (data.projects.length === 0) {
                projectsList.innerHTML = '<div style="color:var(--text-muted);font-size:0.85rem;padding:8px;">暂无保存的项目</div>';
                return;
            }

            const taskNames = {
                'text2video': '✍️', 'image2video': '🖼️',
                'video_edit': '✂️', 'video_extend': '⏩',
            };

            projectsList.innerHTML = data.projects.map(p => `
                <div class="project-item" data-id="${p.id}">
                    <div>
                        <div class="project-name">${taskNames[p.task_type] || ''} ${p.name}</div>
                        <div class="project-meta">${utils.formatDate(p.updated_at)}</div>
                    </div>
                    <div class="project-actions">
                        <button class="btn-icon btn-load-proj" data-id="${p.id}" title="加载">📂</button>
                        <button class="btn-icon btn-del-proj" data-id="${p.id}" title="删除">🗑️</button>
                    </div>
                </div>
            `).join('');

            // 绑定事件
            projectsList.querySelectorAll('.btn-load-proj').forEach(btn => {
                btn.addEventListener('click', (e) => {
                    e.stopPropagation();
                    loadProject(btn.dataset.id);
                });
            });

            projectsList.querySelectorAll('.btn-del-proj').forEach(btn => {
                btn.addEventListener('click', async (e) => {
                    e.stopPropagation();
                    if (confirm('确认删除此项目？')) {
                        await api.delete(`/api/prompts/${btn.dataset.id}`);
                        Toast.success('项目已删除');
                        loadProjectsList();
                    }
                });
            });

        } catch (err) {
            Toast.error(`加载项目列表失败: ${err.message}`);
        }
    }

    async function loadProject(projectId) {
        try {
            const data = await api.get(`/api/prompts/${projectId}`);

            // 还原五要素
            ['subject', 'scene', 'action', 'camera', 'atmosphere'].forEach(key => {
                state[key] = data[key] || '';
                inputs[key].value = state[key];
            });

            // 还原参数
            state.currentProjectId = data.id;
            state.name = data.name;
            state.task_type = data.task_type || 'text2video';
            state.model = data.model || 'doubao-seedance-2-0-260128';
            state.resolution = data.resolution || '720p';
            state.ratio = data.ratio || '16:9';
            state.duration = data.duration || 5;
            state.ref_assets = data.ref_assets || [];

            params.taskType.value = state.task_type;
            params.model.value = state.model;
            params.duration.value = state.duration;
            params.durationDisplay.textContent = state.duration;

            document.querySelectorAll('#param-resolution .chip').forEach(c => {
                c.classList.toggle('active', c.dataset.value === state.resolution);
            });
            document.querySelectorAll('#param-ratio .chip').forEach(c => {
                c.classList.toggle('active', c.dataset.value === state.ratio);
            });

            updatePreview();
            Toast.success(`已加载项目: ${data.name}`);
        } catch (err) {
            Toast.error(`加载项目失败: ${err.message}`);
        }
    }

    // ── 复制与导出 ───────────────────────────────────
    document.getElementById('btn-copy-prompt').addEventListener('click', () => {
        const text = preview.text.textContent;
        if (!text || text.includes('在左侧填写')) {
            Toast.info('请先填写 Prompt 内容');
            return;
        }
        utils.copyToClipboard(text);
    });

    document.getElementById('btn-export-json').addEventListener('click', () => {
        const jsonText = preview.json.textContent;
        try {
            const payload = JSON.parse(jsonText);
            const filename = `seedance_prompt_${state.currentProjectId || 'draft'}.json`;
            utils.downloadJSON(payload, filename);
            Toast.success('JSON 已导出');
        } catch {
            Toast.error('导出失败');
        }
    });

    // ── 初始化预览 ───────────────────────────────────
    updatePreview();
});
